[ZeRO] Add ctor support for multiple param groups

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#72578 [ZeRO] Add ctor support for multiple param groups**

**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

This modifies `test_collect_shards()` to skip if ROCm.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`) versus a non-sharded optimizer.

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

Commit:
[ZeRO] Add ctor support for multiple param groups

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34106940).

Commit:
Update on "[ZeRO] Add ctor support for multiple param groups"


**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`).

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34106940).

Commit:
Update on "[ZeRO] Add ctor support for multiple param groups"


**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`).

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34106940).

rohan-varma commented on test/distributed/optim/test_zero_redundancy_optimizer.py:
```diff
@@ -459,6 +483,55 @@ def some_trainable():
+                loss.backward()
+                optim.step()
+
+            for i in range(len(model1)):
```
It can be made clearer by

```
for lin1, lin2 in zip(model1, model2):
    assert lin1.weight == lin2.weight
    # etc
```

rohan-varma commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -1286,36 +1296,57 @@ def _build_ddp_param_buckets(self) -> None:
+            all_dicts &= isinstance(param, dict)
+        if not all_tensors and not all_dicts:
+            raise TypeError("`params` argument should be an iterable of "
+                            "Tensors or dicts")
```
an easier way to do this might be 

```
if not all(isinstance(tensor, param) for param in all_params) or not all(isinstance(param, dict) for param in all_params):
    # raise 
```

awgu replied:
minor: I actually had something like that originally but decided to make only a single pass over `all_params`. Either way should be fine.

rohan-varma reviewed (commented):
Looks great, thanks for adding this! Just a few minor comments and then will stamp

awgu commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -1286,36 +1296,57 @@ def _build_ddp_param_buckets(self) -> None:
+            # `all_params` contains parameter groups (not parameters)
+            for param_group in all_params:
+                self._all_params.extend(param_group["params"])
+            return params
```
I agree that this is confusing, but hopefully the docstring clarifies things (see the `Returns`).

In the current design, the input is treated as possibly malformed upon entering this method, and `ZeRO` performs its own validation before forwarding arguments to its parent `Optimizer` constructor later in `__init__()`.

The return value of this method is meant to be exactly what to forward into the parent `Optimizer` constructor. The rest of the `ZeRO` implementation relies on `self._all_params` representing a `list` of all parameters (and not any kind of `dict`), so I wanted to preserve that meaning of `self._all_params`.

Now, if `params` is an iterable of `dict`s, then we should maintain that when passing to the parent `Optimizer` constructor to correctly construct `self.param_groups`, so we return `params`. On the other hand, if `params` is an iterable of `Tensor`s, then it may be a generator, in which case it gets exhausted during the validation --- there are some tricky Python semantics at play here, and I could not determine a better solution. Python does not support cloning generators. This is why we return `self._all_params`, which is a `list` and thus can be iterated over again.

awgu replied:
Update: I realized that if somehow `params` is a generator of `dict`s, the same issue holds, so I have updated the logic to handle that case (which probably never happens in practice but is still permitted under the API definition).

This _should_ make things slightly less confusing as well since we always return `all_params` now.

awgu commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -1286,36 +1296,57 @@ def _build_ddp_param_buckets(self) -> None:
+            self._all_params = []
+            # `all_params` contains parameter groups (not parameters)
+            for param_group in all_params:
+                self._all_params.extend(param_group["params"])
```
Added this.

awgu commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -1331,6 +1362,9 @@ def _verify_same_dense_param_type(self) -> None:
         NOTE: This method can be removed once support for sparse parameters
         and varying parameter types is added.
         """
+        assert isinstance(self._all_params[0], torch.Tensor), \
```
Removed this assert since it was for internal debugging. It is guaranteed from `_verify_and_init_params()` that this `assert` will never fail.

awgu commented on test/distributed/optim/test_zero_redundancy_optimizer.py:
```diff
@@ -459,6 +483,55 @@ def some_trainable():
+            {"params": [l.bias for l in model2], "weight_decay": 0.01}
+        )
+
+        # Check parity over a few iterations
```
Good idea.

Commit:
Update on "[ZeRO] Add ctor support for multiple param groups"


**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`).

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34106940).

awgu commented on torch/distributed/optim/zero_redundancy_optimizer.py:
```diff
@@ -1286,36 +1296,57 @@ def _build_ddp_param_buckets(self) -> None:
+            self._all_params = all_params
+            return self._all_params
+        elif all_dicts:
+            self._all_params = []
```
Opted to not go for this since I added a check for `"params" in param_group` while iterating.

Commit:
Update on "[ZeRO] Add ctor support for multiple param groups"


**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`) versus a non-sharded optimizer.

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34106940).

Commit:
Update on "[ZeRO] Add ctor support for multiple param groups"


**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`) versus a non-sharded optimizer.

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34106940).

rohan-varma reviewed (approved):
LGTM, thanks for the quick addition! If you feel like this feature is worth including in the zero tutorial please feel free to make a PR there as well.

Commit:
Update on "[ZeRO] Add ctor support for multiple param groups"


**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`) versus a non-sharded optimizer.

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34106940).

Commit:
Update on "[ZeRO] Add ctor support for multiple param groups"


**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`) versus a non-sharded optimizer.

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34106940).

Commit:
Update on "[ZeRO] Add ctor support for multiple param groups"


**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`) versus a non-sharded optimizer.

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

[ghstack-poisoned]

Commit:
Update on "[ZeRO] Add ctor support for multiple param groups"


**Overview**
This adds `ZeroRedundancyOptimizer` constructor support for multiple parameter groups (i.e. passing an `iterable` of `dict`s instead of an `iterable` of `torch.Tensor` as the `parameters` argument) to mirror the API for non-sharded optimizers.

Fixes https://github.com/pytorch/pytorch/issues/71347 and https://github.com/pytorch/pytorch/issues/59973.

This modifies `test_collect_shards()` to skip if ROCm.

**Test Plan**
I adjusted the existing constructor test, and I added a test for parity between constructing with two parameter groups up front versus constructor with one parameter group and adding the second parameter group after (via `add_param_group()`) versus a non-sharded optimizer.

Differential Revision: [D34106940](https://our.internmc.facebook.com/intern/diff/D34106940)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34106940).

malfet commented:
Reverting as it broke Windows CI, see https://hud.pytorch.org/commit/pytorch/pytorch/5dd07324578f5110a2ec5c213fb559bc49004c7a
